pilot/pkg/networking/plugin/authn/authentication.go

```diff
@@ -41,9 +50,114 @@ func NewPlugin() *Plugin {
+func RequireTLS(policy *authn.Policy) (bool, *authn.MutualTls) {
+	if policy == nil {
+		return false, nil
+	}
+	if len(policy.Peers) > 0 {
+		for _, method := range policy.Peers {
+			switch method.GetParams().(type) {
+			case *authn.PeerAuthenticationMethod_Mtls:
+				return true, method.GetMtls()
+			default:
+				continue
+			}
+		}
+	}
+	return false, nil
+}
+
+// JwksURIClusterName returns cluster name for the jwks URI. This should be used
+// to override the name for outbound cluster that are added for Jwks URI so that they
+// can be referred correctly in the JWT filter config.
+func JwksURIClusterName(hostname string, port *model.Port) string {
+	const clusterPrefix = "jwks."
+	const maxClusterNameLength = 189 - len(clusterPrefix)
+	name := hostname + "|" + port.Name
+	if len(name) > maxClusterNameLength {
+		prefix := name[:maxClusterNameLength-sha1.Size*2]
+		sum := sha1.Sum([]byte(name))
+		name = fmt.Sprintf("%s%x", prefix, sum)
+	}
+	return clusterPrefix + name
+}
+
+// CollectJwtSpecs returns a list of all JWT specs (ponters) defined the policy. This
+// provides a convenient way to iterate all Jwt specs.
+func CollectJwtSpecs(policy *authn.Policy) []*authn.Jwt {
+	ret := []*authn.Jwt{}
+	if policy == nil {
+		return ret
+	}
+	for _, method := range policy.Peers {
+		switch method.GetParams().(type) {
+		case *authn.PeerAuthenticationMethod_Jwt:
+			ret = append(ret, method.GetJwt())
+		}
+	}
+	for _, method := range policy.Origins {
+		ret = append(ret, method.Jwt)
+	}
+	return ret
+}
+
+// OutputLocationForJwtIssuer returns the header location that should be used to output payload if
+// authentication succeeds.
+func OutputLocationForJwtIssuer(issuer string) string {
+	const locationPrefix = "istio-sec-"
+	sum := sha1.Sum([]byte(issuer))
+	return locationPrefix + fmt.Sprintf("%x", sum)
+}
+
+// ConvertPolicyToJwtConfig converts policy into Jwt filter config for envoy.
+func ConvertPolicyToJwtConfig(policy *authn.Policy) *jwtfilter.JwtAuthentication {
+	policyJwts := CollectJwtSpecs(policy)
+	if len(policyJwts) == 0 {
+		return nil
+	}
+	ret := &jwtfilter.JwtAuthentication{
+		AllowMissingOrFailed: true,
+	}
+	for _, policyJwt := range policyJwts {
+		hostname, port, _, err := model.ParseJwksURI(policyJwt.JwksUri)
+		if err != nil {
+			log.Errorf("Cannot parse jwks_uri %q: %v", policyJwt.JwksUri, err)
+			continue
+		}
+
+		jwt := &jwtfilter.JwtRule{
+			Issuer:    policyJwt.Issuer,
+			Audiences: policyJwt.Audiences,
+			JwksSourceSpecifier: &jwtfilter.JwtRule_RemoteJwks{
+				RemoteJwks: &jwtfilter.RemoteJwks{
+					HttpUri: &core.HttpUri{
+						Uri: policyJwt.JwksUri,
+						HttpUpstreamType: &core.HttpUri_Cluster{
+							Cluster: JwksURIClusterName(hostname, port),
+						},
+					},
+					CacheDuration: &types.Duration{Seconds: jwtPublicKeyCacheSeconds},
+				},
+			},
+			ForwardPayloadHeader: OutputLocationForJwtIssuer(policyJwt.Issuer),
+			Forward:              true,
```

Comment:
why do you want to forward the token?

Reply:
Just to follow the old code. We can remove if there is no need to keeping it. Prefer to do it in different PR though.

Reply:
I feel it is important not to forward the token.   Please file an issue to track it

Reply:
Ok. I've changed it to false in this PR.